Modules/_multiprocessing/posixshmem.c

```diff
@@ -0,0 +1,724 @@
+            break;
+
+            case EEXIST:
+                PyErr_SetString(pExistentialException,
+                                "Shared memory with the specified name already exists");
+            break;
+
+            case ENOENT:
+                PyErr_SetString(pExistentialException,
+                                "No shared memory exists with the specified name");
+            break;
+
+            case EINVAL:
+                PyErr_SetString(PyExc_ValueError, "Invalid parameter(s)");
+            break;
+
+            case EMFILE:
+                PyErr_SetString(PyExc_OSError,
+                                 "This process already has the maximum number of files open");
+            break;
+
+            case ENFILE:
+                PyErr_SetString(PyExc_OSError,
+                                 "The system limit on the total number of open files has been reached");
+            break;
+
+            case ENAMETOOLONG:
+                PyErr_SetString(PyExc_ValueError,
+                                 "The name is too long");
+            break;
+
+            default:
+                PyErr_SetFromErrno(PyExc_OSError);
+            break;
+        }
+
+        goto error_return;
+    }
+    else {
+        if (size) {
+            DPRINTF("calling ftruncate, fd = %d, size = %ld\n", self->fd, size);
+            if (-1 == ftruncate(self->fd, (off_t)size)) {
+                // The code below will raise a Python error. Since that error
+                // is raised during __init__(), it will look to the caller
+                // as if object creation failed entirely. Here I clean up
+                // the system object I just created.
+                close(self->fd);
+                shm_unlink(self->name);
+
+                // ftruncate can return a ton of different errors, but most
+                // are not relevant or are extremely unlikely.
+                switch (errno) {
+                    case EINVAL:
+                        PyErr_SetString(PyExc_ValueError,
+                                        "The size is invalid or the memory is read-only");
+                    break;
+
+                    case EFBIG:
+                        PyErr_SetString(PyExc_ValueError,
+                                        "The size is too large");
+                    break;
+
+                    case EROFS:
+                    case EACCES:
+                        PyErr_SetString(pPermissionsException,
+                                        "The memory is read-only");
+                    break;
+
+                    default:
+                        PyErr_SetFromErrno(PyExc_OSError);
+                    break;
+                }
+
+                goto error_return;
+            }
+        }
+    }
+
+    return 0;
+
+    error_return:
+    return -1;
+}
+
+
+static void SharedMemory_dealloc(SharedMemory *self) {
+    DPRINTF("dealloc\n");
+    PyMem_Free(self->name);
+    self->name = NULL;
```

Comment:
Should this not clean up the file descriptor in self->fd?

Reply:
Removed in GH-11816.